### PR TITLE
add token based pagination for list workspaces endpoint

### PR DESCRIPTION
**Pull request by linetdsouza**
Add Token-Based Pagination Support for ListWorkspaces Endpoint

  Summary

Implements token-based pagination for the WorkspaceResources.ListWorkspaces() method to complement the existing offset-based pagination. This enhancement provides more efficient pagination for large workspaces and aligns with modern API pagination patterns.

  Changes Made

  Core Implementation

  - New Models: Added TokenPaginationParameters and TokenPaginatedResult<T> classes to support token-based pagination
  - API Enhancement: Extended WorkspaceResources interface with new ListWorkspaces(TokenPaginationParameters) overload
  - Implementation: Updated WorkspaceResourcesImpl to handle token pagination requests and responses
  - JSON Support: Enhanced serializers (JsonNetSerializer, JsonSerializer) to properly handle new pagination models

  Key Features

  - Token Pagination: Support for lastKey token to retrieve subsequent pages
  - Flexible Parameters: Support for maxItems and paginationType configuration
  - Backward Compatibility: Existing offset-based pagination remains unchanged

  Testing

    - Comprehensive Test Coverage: Added ListWorkspacesWithTokenPagination test method with multiple scenarios:
    - Basic token pagination with maxItems parameter
    - Explicit paginationType parameter validation
    - Multi-page navigation using lastKey tokens
    - Edge case handling for invalid parameters using Assert.ThrowsException
    - Mock API Tests: Added TokenPaginationParametersTest for unit testing pagination parameters

  Documentation

  - CHANGELOG: Updated with version 3.1.0 release notes
  - API Documentation: Added comprehensive XML documentation for new methods and classes

To test the SDK package change manually:

```
- export SMARTSHEET_ACCESS_TOKEN=xxxxxxxxxxxxx
- cd sdk-csharp-sample-net80
- dotnet run

Smartsheet C# SDK Examples
==========================

Choose an example to run:
1. Simple Example (Basic sheet operations)
2. Token Pagination Example (New workspace pagination features)
3. HTTP Client Example

Enter your choice (1-3): 2

Running Token Pagination Example...

Token Pagination Example - Smartsheet API C# SDK
=================================================

1. Demonstrating Token-Based Pagination for Workspaces
======================================================
Fetching workspaces with token pagination (maxItems=100)...

✓ Retrieved 2 workspaces
✓ Last Key: bGFzdEtleT1leUpzWVhOMFJYWmhiSFZoZEdWa1MyVjVJam9pSWl3aWJHRnpkRk55YmlJNkluVnpaWEl1TmpJek5EQTBOemdpTENKbGJuVnRaWEpoZEdsdmJrRmpZMlZ6YzFOMFlYSjBVRzlwYm5SbGNpSTZJa1JKVWtWRFZDSXNJbXhoYzNSRmRtRnNkV0YwWldSTFpYbHp
✓ Pagination Type: token-based

Workspace Details:
------------------
  • ID: 111111, Name: 'xxxxxx'
  • ID: 22222, Name: 'yyyyy'


2. Demonstrating Multi-Page Token Pagination
============================================
Demonstrating multi-page token pagination (maxItems=100)...

Page 1:
  ├─ Retrieved: 2 workspaces
  ├─ Last Key: bGFzdEtleT1leUpzWVhOMFJYWmhiSFZoZEdWa1MyVjVJam9pSWl3aWJHRnpkRk55YmlJNkluVnpaWEl1TmpJek5EQTBOemdpTENKbGJuVnRaWEpoZEdsdmJrRmpZMlZ6YzFOMFlYSjBVRzlwYm5SbGNpSTZJa1JKVWtWRFZDSXNJbXhoYzNSRmRtRnNkV0YwWldSTFpYbHpJanA3S
  │   • xxx (ID: 1111)
  │   • yyyy (ID: 2222)
  └─ Total so far: 2 workspaces

Page 2:
  ├─ Retrieved: 8 workspaces
  ├─ Last Key: bGFzdEtleT1leUpzWVhOMFJYWmhiSFZoZEdWa1MyVjVJam9pSWl3aWJHRnpkRk55YmlJNkluVnpaWEl1TmpJek5EQTBOemdpTENKbGJuVnRaWEpoZEdsdmJrRmpZMlZ6YzFOMFlYSjBVRzlwYm5SbGNpSTZJa1JKVWtWRFZDSXNJbXhoYzNSRmRtRnNkV0YwWldSTFpYbHpJanA3
  │   • x1 (ID: 1111)
  │   • x2 (ID: 2222)
  │   • x3 (ID: 333)
  │   • x4  (ID: 44)
  │   • x5 (ID: 555)
  │   • x6 (ID: 666)
  │   • x7(ID: 777)
  │   • x8 (ID: 888)
  └─ Total so far: 10 workspaces

Page 3:
  ├─ Retrieved: 2 workspaces
  ├─ Last Key: bGFzdEtleT1leUpzWVhOMFJYWmhiSFZoZEdWa1MyVjVJam9pSWl3aWJHRnpkRk55YmlJNkluVnpaWEl1TmpJek5EQTBOemdpTENKbGJuVnRaWEpoZEdsdmJrRmpZMlZ6YzFOMFlYSjBVRzlwYm5SbGNpSTZJa1JKVWtWRFZDSXNJbXhoYzNSRmRtRnNkV0YwWldS
  │   • xxxxx(ID: 34343)
  │   • YYYYY (ID: 34343434)
  └─ Total so far: 12 workspaces

✓ Total workspaces found: 12
✓ Pages processed: 3

3. Comparing Traditional vs Token Pagination
============================================
Comparing traditional offset vs token-based pagination...

Traditional Pagination (PaginationParameters):
  ├─ Type: Offset-based
  ├─ Page Size: 100
  ├─ Retrieved: 100 workspaces
  ├─ Total Count: 784
  └─ Has More: True

Token-Based Pagination (TokenPaginationParameters):
  ├─ Type: Token-based
  ├─ Max Items: 100
  ├─ Retrieved: 2 workspaces
  ├─ Last Key: bGFzdEtleT1leUpzWVhOMFJYWmhiSFZoZEdWa1MyVjVJam9pSWl3aWJHRnpkRk55YmlJNkluVnpaWEl1TmpJek5EQTBOemdpTENKbGJuVnRaWEpoZEdsdmJrRmpZMlZ6YzFOMFlYSjBVRzlwYm5SbGNpSTZJa1JKVWtWRFZDSXNJbXhoYzNSRmRtRnNkV0YwWldSTFpYbH
  └─ Has More: True

```

📝 Key Differences:
   • Traditional: Uses offset/limit, provides total count
   • Token-based: Uses lastKey, more efficient for large datasets
   • Token approach prevents issues with data changes during pagination

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,9 +3,14 @@ All notable changes to this project will be documented in this file.
...
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
-
 ## [Unreleased]
 
+## [3.1.0] - 2025-08-12
+### Added
+- Add support for token-based pagination in WorkspaceResources.ListWorkspaces()
+- New TokenPaginationParameters class to support paginationType, lastKey and maxItems parameters
+- New TokenPaginatedResult<T> class for token-based pagination responses with data and lastKey properties
+
```
Comment (smar-tim-hodges) [inline CHANGELOG.md]: The latest tagged version is 5.0.0 https://github.com/smartsheet/smartsheet-csharp-sdk/releases/tag/5.0.0

You can also see how to update the version in this PR. https://github.com/smartsheet/smartsheet-csharp-sdk/pull/91/files

File: mock-api-test-sdk-net80/TokenPaginationParametersTest.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Smartsheet.Api.Models;
+using System.Collections.Generic;
+
+namespace mock_api_test_sdk_net80
+{
+    [TestClass]
+    public class TokenPaginationParametersTest
```
Comment (smar-brian-keare) [inline mock-api-test-sdk-net80/TokenPaginationParametersTest.cs:8]: This file appears specific to ListWorkspaces. Maybe consider updating the class name to reflect that.

File: CHANGELOG.md (hunk 1 above)
Comment (linetdsouza) [reply CHANGELOG.md]: Thanks. I have updated it.

**Review by smar-tim-hodges — APPROVED** [review]
Great job on this and am super glad to see external work done for the SDK outside of maintainers.

File: smartsheet-csharp-sdk/main/Smartsheet/Api/Models/TokenPaginationParameters.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
+using Smartsheet.Api.Internal.Util;
+using System.Collections.Generic;
+
+namespace Smartsheet.Api.Models
+{
+    /// <summary>
+    /// Builds optional query string parameters for token-based pagination.
+    /// </summary>
+    public class TokenPaginationParameters
```
Comment (smar-brian-keare) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/Models/TokenPaginationParameters.cs:9]: Would the intention be for other endpoints that use lastKey and maxItems to use this class? Should there be a base implementation with just lastKey and maxItems that could be extended with more params like paginationType?

File: mock-api-test-sdk-net80/TokenPaginationParametersTest.cs (hunk 2 above)
Comment (linetdsouza) [reply mock-api-test-sdk-net80/TokenPaginationParametersTest.cs:8]: This can be used for any enumeration list endpoint. List Workspaces is the first endpoint using TokenPaginationParameters.

File: smartsheet-csharp-sdk/main/Smartsheet/Api/Models/TokenPaginationParameters.cs (hunk 3 above)
Comment (linetdsouza) [reply smartsheet-csharp-sdk/main/Smartsheet/Api/Models/TokenPaginationParameters.cs:9]: Yes, we can create base pagination parameters for any future implementation where we need only maxItems and lastKey.
